types/underscore/index.d.ts

```diff
@@ -97,11 +97,12 @@ declare module _ {
 
     type EnumerableKey = string | number;
 
-    type CollectionIterator<T, TResult, V> =
-        V extends List<T> ? ListIterator<T, TResult, V>
-        : V extends Dictionary<T> ? ObjectIterator<T, TResult, V>
```

Comment:
To provide some context here, I ran into two issues in my company's solution where I was ending up with a union of iterators that caused type inference to become unhappy that prompted this change. One was a `File[] | FileList` (which resulted in a union of two possible iterators with different `V`s) and the other was `any` (which resulted in a union of a `ListIterator` and `ObjectIterator`). Since the element and collection types don't vary in those cases, I updated this type to specifically vary the `key` parameter based on whether or not the collection union included a list or dictionary (or `any`, which could provide either type of key).

Reply:
I guess, too, this highlights one of the issues with the tests that I've written - I currently declare function iteratees at the top of the file and refer to them, but because I have to declare parameter types when I do that that practice can end up masking parameter type inference issues like this one.

Perhaps I should add a few tests that test a few of the types themselves such as the below.

```TS
declare const unionCollectionIteratee: Iteratee<StringRecord[] | _.Dictionary<StringRecord>, void> = (element, key, collection) => {
    element; // $ExpectType StringRecord
    key; // $ExpectType string | number
    collection; // $ExpectType StringRecord[] | _.Dictionary<StringRecord>
};
```

Reply:
Well-caught and very elegant solution. Nicely in line with `IterateeResult` as well.

I agree tests like your example would be useful.

Reply:
Super extra good news! I applied this to `MemoCollectionIterator` as well and that resolved that weird type narrowing issue for constant initial aggregate values that kept me from collapsing `UnderscoreStatic`'s `reduce` and `reduceRight` down to one overload each 😁

Reply:
I've added tests like those I described for `Iteratee` and `IterateeResult` since they're complicated, let me know what you think.